src/cls/rgw/cls_rgw.cc

```diff
@@ -2656,7 +2656,9 @@ static int bi_log_iterate_entries(cls_method_context_t hctx, const string& marke
       if (ret < 0)
         return ret;
       i++;
-
+    }
+    if (iter == keys.end()) {
+      break;
```

Comment:
the bz and commit message talk about a segfault in `gc_iterate_entries()`, but this modifies `bi_log_iterate_entries()`

i'm also not sure this change is correct. it looks like `iter == keys.end()` will always be the case here, because the `for (; iter != keys.end(); ++iter)` loop above doesn't have any breaks. so breaking here would prevent us from continuing the listing

also, it's perfectly safe to decrement the end iterator - this will give you the last entry as long as the container isn't empty - and we check that `begin() != end()` before the for-loop

it looks like the segfault is instead coming from [this early return](https://github.com/ceph/ceph/blob/fb6ebfb1937b1880468f0806719668459608e041/src/cls/rgw/cls_rgw.cc#L3270-L3274) in gc_iterate_entries(), where `iter` could still be at `begin()`. i think this should fix?
```diff
diff --git a/src/cls/rgw/cls_rgw.cc b/src/cls/rgw/cls_rgw.cc
index ff994dc5c6..aa870df3e3 100644
--- a/src/cls/rgw/cls_rgw.cc
+++ b/src/cls/rgw/cls_rgw.cc
@@ -3268,8 +3268,12 @@ static int gc_iterate_entries(cls_method_context_t hctx, const string& marker, b
       if (max_entries && (i >= max_entries)) {
         if (truncated)
           *truncated = true;
-        --iter;
-        key_iter = iter->first;
+        if (iter == keys.begin()) {
+          key_iter = start_key;
+        } else {
+          --iter;
+          key_iter = iter->first;
+        }
         return 0;
       }
```

Reply:
sorry, I'll close this but need a fix in place todya